### PR TITLE
Direction tool

### PR DESCRIPTION
**Pull request by ivanbusthomi**
Work in progress

### What does it fix?
* Ticket: #4262
* Funded by: 
* Description: Added direction and distance tool

### Checklist:
- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Add to the changelog in metadata.txt if it's a new feature
- [ ] Unit test for new code added
- [ ] Request someone to review or test your PR

### CODE REVIEW

File: safe/gui/tools/direction_tool.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    QgsFeature,
+    QgsGeometry,
+    QgsField,
+    QgsCoordinateReferenceSystem,
+    QgsCoordinateTransform)
+import math
+
+FORM_CLASS = get_ui_class('direction_tool_base.ui')
+
+class DirectionTool(QtGui.QDialog, FORM_CLASS):
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E302] expected 2 blank lines, found 1

File: safe/gui/tools/direction_tool.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        direction_list = [
+            'N', 'NNE', 'NE', 'ENE', 'E', 'ESE', 'SE',
+            'SSE', 'S', 'SSW', 'SW', 'WSW', 'W', 'WW',
+            'NW', 'NNW'
+        ]
+
+        try:
+            bearing = float(bearing_angle)
+        except ValueError:
+            print 'Error casting bearing to a float'
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E999] SyntaxError: invalid syntax

Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E999] SyntaxError: invalid syntax

File: safe/gui/tools/direction_tool.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+        direction_count = len(direction_list)
+        direction_interval = 360. / direction_count
+        index = int(round(bearing / direction_interval))
+        index %= direction_count
+        return direction_list[index]
+
+
+    def distance_and_bearing(self, hazard_layer, exposure_layer, update_layer=True):
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E303] too many blank lines (2)
* [E501] line too long (84 > 79 characters)

Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E303] too many blank lines (2)
* [E501] line too long (84 > 79 characters)

File: safe/gui/tools/direction_tool.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        bearing_index = index + 1
+        direction_index = index + 2
+
+        hazard_feature_list = []
+        for feature in hazard_layer.getFeatures():
+            hazard_feature_list.append(feature)
+        if len(hazard_feature_list) <= 1:
+            hazard_feature = hazard_feature_list[0]
+        else:
+            print 'hazard layer contain more than one feature'  # implement logging later
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (89 > 79 characters)

Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (89 > 79 characters)

File: safe/gui/tools/direction_tool.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            print 'using the first feature as default'          # implement logging later
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (89 > 79 characters)

Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (89 > 79 characters)

File: safe/gui/tools/direction_tool.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        hazard_point = hazard_feature.geometry().asPoint()
+        exposure_layer.startEditing()
+        for feature in exposure_layer.getFeatures():
+            fid = feature.id()
+            exposure_point = feature.geometry().asPoint()
+            # calculate cardinality
+            bearing_angle = hazard_point.azimuth(exposure_point)
+            cardinality = self.bearing_to_cardinal(bearing_angle)
+            # calculate distance
+            measure_distance = distance.measureLine(hazard_point, exposure_point)
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (81 > 79 characters)

Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (81 > 79 characters)

File: safe/gui/tools/direction_tool.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+            exposure_layer.changeAttributeValue(fid, distance_index, measure_distance)
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (86 > 79 characters)

Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (86 > 79 characters)

File: safe/gui/tools/direction_tool.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            exposure_layer.changeAttributeValue(fid, bearing_index, bearing_angle)
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (82 > 79 characters)

Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (82 > 79 characters)

File: safe/gui/tools/direction_tool.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            exposure_layer.changeAttributeValue(fid, direction_index, cardinality)
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (82 > 79 characters)

Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: __[flake8]__
* [E501] line too long (82 > 79 characters)

File: safe/gui/tools/direction_tool.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
+from PyQt4 import QtGui
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: Add header files. See the example in the other files.

File: safe/gui/tools/direction_tool.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+        self.hazard.setFilters(QgsMapLayerProxyModel.PointLayer)
+        self.exposure.setFilters(QgsMapLayerProxyModel.PointLayer)
+
+        ok_button = self.buttonBox.button(QtGui.QDialogButtonBox.Ok)
+        ok_button.clicked.connect(self.accept)
+        cancel_button = self.buttonBox.button(QtGui.QDialogButtonBox.Cancel)
+        cancel_button.clicked.connect(self.reject)
+
+    def bearing_to_cardinal(self, bearing_angle):
```
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: Method's docstring

File: safe/gui/tools/direction_tool.py (hunk 2 above)
Comment (ismailsunni) [inline safe/gui/tools/direction_tool.py]: In general, avoid to use print.
For logging, you can do somethting like this:
```
import logging
LOGGER = logging.getLogger('InaSAFE')

# your code
LOGGER.debug('It works!!!')

**Comment by Gustry** [comment]
Hi @ivanbusthomi ,

Can you update your branch from develop?
We added some checks on the develop branch about python. Thanks

**Comment by ivanbusthomi** [comment]
Following our discussion in the last site visit, I've moved the changes to a new pull request #4507

**Pull request closed without merging**